sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs

```diff
@@ -153,6 +188,81 @@ private async ValueTask<bool> AuthorizeRequestOnChallengeAsyncInternal(HttpMessa
             return true;
         }
 
+        /// <inheritdoc />
+        public override ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            return ProcessAsyncInternal(message, pipeline, true);
+        }
+
+        /// <inheritdoc />
+        public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessAsyncInternal(message, pipeline, false).EnsureCompleted();
+        }
+
+        private async ValueTask ProcessAsyncInternal(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
+        {
+            if (message.Request.Uri.Scheme != Uri.UriSchemeHttps)
+            {
+                throw new InvalidOperationException("Bearer token authentication is not permitted for non TLS protected (https) endpoints.");
+            }
+
+            if (async)
+            {
+                await AuthorizeRequestAsync(message).ConfigureAwait(false);
+                await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+            }
+            else
+            {
+                AuthorizeRequest(message);
+                ProcessNext(message, pipeline);
+            }
+
+            // Check if we have received a challenge or we have not yet issued the first request.
+            if (message.Response.Status == (int)HttpStatusCode.Unauthorized && message.Response.Headers.Contains(HttpHeader.Names.WwwAuthenticate))
+            {
+                // Attempt to get the TokenRequestContext based on the challenge.
+                // If we fail to get the context, the challenge was not present or invalid.
+                // If we succeed in getting the context, authenticate the request and pass it up the policy chain.
+                if (async)
+                {
+                    if (await AuthorizeRequestOnChallengeAsync(message).ConfigureAwait(false))
+                    {
+                        await ProcessNextAsync(message, pipeline).ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    if (AuthorizeRequestOnChallenge(message))
+                    {
+                        ProcessNext(message, pipeline);
+                    }
+                }
+            }
+
+            // Handle the scenario in which we get a CAE challenge back.
+            if (message.Response.Status == (int)HttpStatusCode.Unauthorized
```

Comment:
should this if block be nested inside the previous one above? My understanding is that we only want to check for a second challenge if the first check above processed a challenge.

Reply:
True, it's a retry so it should be inside the first `if` block.